plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitManager.java

```diff
@@ -108,6 +112,31 @@ public ConnectorSplitSource getSplits(
         return new ClassLoaderSafeConnectorSplitSource(splitSource, IcebergSplitManager.class.getClassLoader());
     }
 
+    private Scan<?, FileScanTask, CombinedScanTask> getScan(IcebergMetadata icebergMetadata, Table icebergTable, IcebergTableHandle table, ExecutorService executor)
+    {
+        Long fromSnapshot = icebergMetadata.getIncrementalRefreshFromSnapshot(table.getSchemaTableName()).orElse(null);
+        if (fromSnapshot != null) {
+            // check if fromSnapshot is still part of the table's snapshot history
+            boolean containsFromSnapshot = SnapshotUtil.isAncestorOf(icebergTable, fromSnapshot);
+            if (containsFromSnapshot) {
+                boolean containsModifiedRows = false;
+                for (Snapshot snapshot : SnapshotUtil.ancestorsBetween(icebergTable, icebergTable.currentSnapshot().snapshotId(), fromSnapshot)) {
+                    if (snapshot.operation().equals(DataOperations.OVERWRITE) || snapshot.operation().equals(DataOperations.DELETE)) {
+                        containsModifiedRows = true;
+                        break;
+                    }
+                }
+                if (!containsModifiedRows) {
+                    return icebergTable.newIncrementalAppendScan().fromSnapshotExclusive(fromSnapshot).planWith(executor);
+                }
+            }
+            // fromSnapshot is missing (could be due to snapshot expiration or rollback), or snapshot range contains modifications
+            // (deletes or overwrites), so we cannot perform incremental refresh. Falling back to full refresh.
+            icebergMetadata.disableIncrementalRefresh();
```

Comment:
Yeah, good point on this being problematic if there are multiple source tables. That shouldn't be a problem anymore, now that we've restricted the support for single source table MVs.

> I think this is NOOP since right now there can only be one source table handle

For single source table, this method call is still required to clear out the contents of the `fromSnapshotsForRefresh` map. We check that map in `finishRefreshMaterializedView()` to decide whether to truncate the table first (full refresh) or not (incremental refresh): 
```
        boolean isFullRefresh = fromSnapshotsForRefresh.get().isEmpty();
        if (isFullRefresh) {
            // delete before insert .. simulating overwrite
            log.info("Performing full MV refresh for storage table: %s", table.getName());
            transaction.newDelete()
                    .deleteFromRowFilter(Expressions.alwaysTrue())
                    .commit();
        }
        else {
            log.info("Performing incremental MV refresh for storage table: %s", table.getName());
        }
```